arkouda/numeric.py

```diff
@@ -2182,3 +2183,58 @@ def count_nonzero(pda):
         return sum((pda).astype(np.int64))
     elif pda.dtype == str:
         return sum((pda != "").astype(np.int64))
+
+
+def putmask(pda: pdarray, mask: Union[bool, pdarray], values: pdarray):
+    """
+    Overwrite elements of a pdarray at indices where mask is True
+
+    Parameters
+    ----------
+    pda    : pdarray, source data, also output data
+        pda = input where mask is False, = values where mask is True
+    mask   : a scalar boolean, or a pdarray of booleans
+    values : pdarray, replacement data
+
+    Returns
+    -------
+    None - pda is modified in-place
+
+    Notes
+    -----
+    If values.size != a.size, values is repeated and/or pruned as needed to
+    make sizes match, because ak.where requires matching sizes.
+
+    Examples
+    -------
+    >>> a = ak.array(np.arange(10))
+    >>> ak.putmask (a,a>2,a**2)
+    array ([0,1,2,9,16,25,36,49,64,81])
+
+    >>> values = ak.array([3,2])
+    >>> ak.putmask (a,a>2,values)
+    array ([0,1,2,2,3,2,3,2,3,2])
+
+    Raises
+    ------
+    TypeError
+        Raised if a and values are not the same type
+
+    """
+
+    from arkouda.pdarraysetops import concatenate
+
+    # check for matching types
+
+    if values.dtype != pda.dtype:
+        raise TypeError("ak.putmask requires arrays of matching type")
+
+    # if values is not the same size as pda, repeat it and/or prune it as needed
+
+    growth = pda.size // values.size + (0 if pda.size % values.size == 0 else 1)
+    result = concatenate(growth * [values])
+    if result.size > pda.size:
+        reduction = result.size % pda.size
+        result = result[:-(reduction)]
```

Comment:
I feel like this could be more space and time efficient if we did this on the chpl side. Because if `values.size == 2`, we could just loop over these 2 over and over and write them into place instead of allocating `pda.size` elements and concatenating them together.

We could also do some sort of thresholding where if `values` is quite small we make a copy of it on each locale. I'm thinking something similar to how we have 3 different methods for histogram depending on the bucket size. I think the same thing applies here but instead of writing into, we are reading from (so we wouldn't need atomics, we can just use `const`)

Reply:
I'll look into pushing this chapel-side.

Reply:
Had a very good and short conversation with Amanda about this, the upshot of which is that I'm going to look at arkouda-izing np.tiles and np.repeat, and then return to this.

Reply:
we talked about this a bit. I think let's merge this to get some unit tests in it and then we'll do 2 phases of adding more optimized / complex chapel implementations

Reply:
see https://github.com/Bears-R-Us/arkouda/issues/3428 for more info on the future work